Enhancement:GPU-Accelerated Async Video Encoding

linshenghou opened the pull request:
# GPU-Accelerated Async Video Encoding with MJPEG Enhancement

## 🎯 Overview

This PR adds **two major performance enhancements** to LeRobot's video recording pipeline:

1. **GPU-Accelerated Encoding**: 3-4x speedup using NVIDIA NVENC
2. **Asynchronous Processing**: Non-blocking background encoding
3. **MJPEG Enhancement**: Improved camera configuration support

## 🚀 Dual Enhancement Benefits

### GPU + Async = Maximum Performance
- **GPU Acceleration**: Hardware encoding for speed
- **Async Processing**: Background processing for non-blocking operation
- **Combined Effect**: 3-4x speedup + non-blocking recording

## 📊 Performance Results

### Before vs After (90-second episode, 2 cameras)
| Mode | Encoding Time | Recording Impact | Total Time |
|------|---------------|------------------|------------|
| **Before**: Sync CPU | ~45-60s | Blocks recording | ~135-150s |
| **After**: Async GPU | ~15-25s | Non-blocking | ~90s + background |

### Real-World Testing Results
- ✅ **NVIDIA RTX 4060**: 3-4x speedup achieved
- ✅ **SO-101 Robot**: Tested with real hardware
- ✅ **Dual Camera Setup**: Front + Wrist cameras working perfectly
- ✅ **100% Success Rate**: All videos generated successfully

## 🛠️ Usage

### Enable Both Enhancements
```bash
python -m lerobot.record \
--robot.type=so101_follower \
--robot.port=/dev/ttyACM0 \
--robot.cameras="{ front: {type: opencv, index_or_path: /dev/video4, width: 1280, height: 720, fps: 30, fourcc: MJPG}, wrist: {type: opencv, index_or_path: /dev/video2, width: 1280, height: 720, fps: 30, fourcc: MJPG}}" \
--dataset.gpu_video_encoding=true \
--dataset.async_video_encoding=true \
--dataset.video_encoding_workers=2 \
--dataset.episode_time_s=90 \
--dataset.num_episodes=20
```

### Individual Feature Usage
```bash
# GPU only (synchronous)
--dataset.gpu_video_encoding=true
--dataset.async_video_encoding=false

# Async only (CPU)
--dataset.gpu_video_encoding=false
--dataset.async_video_encoding=true
--dataset.video_encoding_workers=2

# Both (recommended)
--dataset.gpu_video_encoding=true
--dataset.async_video_encoding=true
--dataset.video_encoding_workers=2
```

## 🧪 Testing

### Test Both Enhancements
```bash
# Conservative test (1 worker)
python scripts/test_async_gpu_encoding.py --mode conservative

# Aggressive test (2 workers)
python scripts/test_async_gpu_encoding.py --mode aggressive
```

### Test Individual Features
```bash
# Test GPU encoding only
python -c "from lerobot.datasets.gpu_video_encoder import GPUVideoEncoder; print(GPUVideoEncoder().get_encoder_info())"

# Test async encoding only
python -c "from lerobot.datasets.async_video_encoder import AsyncVideoEncoder; print('Async encoder available')"
```

## 📁 Files Added

### Core Implementation
- `src/lerobot/datasets/gpu_video_encoder.py` - GPU encoding implementation
- `src/lerobot/datasets/sync_gpu_encoder.py` - Synchronous GPU encoder
- `src/lerobot/datasets/async_video_encoder.py` - Async encoder with improvements

### Utilities
- `scripts/regenerate_dataset_videos.py` - Video regeneration utility
- `scripts/test_async_gpu_encoding.py` - Comprehensive test script
- `GPU_ENCODING_README.md` - Documentation

## 📝 Files Modified

### Enhanced Core Files
- `src/lerobot/datasets/lerobot_dataset.py` - Added GPU + async support
- `src/lerobot/record.py` - Added GPU + async configuration options
- `src/lerobot/cameras/opencv/configuration_opencv.py` - MJPEG enhancement

## 🔧 Configuration Options

### GPU + Async Configuration
```python
# Enable both enhancements
--dataset.gpu_video_encoding=true
--dataset.async_video_encoding=true
--dataset.video_encoding_workers=2
--dataset.video_encoding_queue_size=100
--dataset.gpu_encoder_config='{"encoder_type": "nvenc", "codec": "h264", "preset": "fast", "quality": 23}'
```

## ✅ Testing Results

### Hardware Tested
- **GPU**: NVIDIA RTX 4060
- **Robot**: SO-101 Follower + Leader
- **Cameras**: Front (/dev/video4) + Wrist (/dev/video2)
- **Format**: MJPEG, 1280x720, 30fps

### Performance Metrics (Async GPU)
- **Episode 0**: 3.87s encoding time (599 frames)
- **Episode 1**: 4.16s encoding time (597 frames)  
- **Episode 2**: 2.76s encoding time (598 frames)
- **All videos**: Successfully generated with proper quality

### Reliability
- ✅ **No stucking issues**: Proper timeout and fallback
- ✅ **Clean shutdown**: All workers stopped correctly
- ✅ **Memory management**: No memory leaks detected
- ✅ **Error handling**: Automatic CPU fallback on GPU failure

## 🎯 Benefits

1. **Dual Performance**: GPU speed + async non-blocking
2. **Flexibility**: Can use GPU only, async only, or both
3. **Reliability**: Automatic fallbacks and error handling
4. **Compatibility**: Works with existing LeRobot workflows
5. **Quality**: Enhanced MJPEG camera support

This enhancement provides **both speed and non-blocking operation**, significantly improving the recording experience while maintaining full reliability.

Commit:
Enhance OpenCVCamera with FOURCC support and validation

- Added FOURCC configuration option to OpenCVCamera and OpenCVCameraConfig for specifying video format.
- Implemented _validate_fourcc method to validate and set the camera's FOURCC code.
- Updated _configure_capture_settings to apply FOURCC settings before FPS and resolution.
- Enhanced camera detection to include default FOURCC code in camera info.
- Updated documentation to reflect new FOURCC parameter and its implications on performance.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Add tests for FOURCC configuration in OpenCVCamera

- Implemented tests to validate FOURCC configuration and its application in OpenCVCamera.
- Added checks for valid FOURCC codes and ensured that invalid codes raise appropriate errors.
- Included a test for camera connection functionality using specified FOURCC settings.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Fix circular import in __init__.py - change to relative import

Commit:
Add comprehensive benchmarking system for encoding speed analysis

Commit:
Fix division by zero issues in benchmarking system

Commit:
Implement async video encoding for improved recording performance

Commit:
Fix AsyncVideoEncoder shutdown and add comprehensive tests

Commit:
Add comprehensive async encoding implementation summary

Commit:
Add async encoding support to benchmarking script with performance comparison

Commit:
Add encoding timing test and monitoring tools for async encoding

Commit:
Fix async video encoder path construction for correct dataset structure

Commit:
Fix async video encoder path construction - use video keys directly without double prefixing

Commit:
Add GPU-accelerated video encoding support with NVIDIA NVENC, Intel QSV, and AMD VCE

Commit:
Add comprehensive GPU encoding guide and test scripts

Commit:
Clarify distinction between async encoding and GPU acceleration as independent features

Commit:
Remove outdated documentation and benchmark scripts related to video encoding, including async and GPU encoding guides, performance measurement tools, and test scripts. This cleanup streamlines the codebase and eliminates redundancy following recent enhancements and optimizations.

Commit:
Add GPU-accelerated video encoding with async support

- GPU acceleration using NVIDIA NVENC (3-4x speedup)
- Async background encoding with configurable workers
- Automatic CPU fallback for reliability
- Timeout protection to prevent stuck processes
- Tested with real robot hardware (SO-101)
- Includes test scripts and concise documentation

Performance: Non-blocking recording with 3-4x encoding speedup

Commit:
Remove outdated GPU encoding issues documentation and related scripts to streamline the codebase following recent enhancements. This cleanup eliminates redundancy and focuses on the latest improvements in video encoding features.

linshenghou committed:
Fix linting and security issues in gpu_video_encoder.py

- Remove unused variable 'result' in encode_video method
- Use shutil.which() to get full path to ffmpeg for security
- Fix bandit B607 warning about partial executable path

linshenghou committed:
Apply code formatting fixes from pre-commit

- Format Python files with ruff
- Fix code style and formatting issues
- Ensure consistent code formatting across the codebase

linshenghou committed:
Apply remaining formatting fixes

- Fix trailing whitespace in GPU_ENCODING_README.md and Python files
- Apply ruff format to lerobot_dataset.py and record.py
- Ensure all files meet pre-commit formatting standards

linshenghou committed:
Apply final formatting fixes from pre-commit

- Fix ruff linting issues (16 errors resolved)
- Apply pyupgrade syntax updates to Python files
- Apply prettier formatting to markdown files
- Ensure all files meet pre-commit standards

linshenghou committed:
Apply final pyupgrade syntax updates

- Update Python syntax to Python 3.9+ standards
- Apply modern Python syntax improvements
- Ensure all files meet latest pyupgrade requirements

linshenghou committed:
Apply final pre-commit fixes

- Fix remaining 10 ruff linting errors
- Apply final pyupgrade syntax updates
- Ensure all files meet pre-commit standards completely

linshenghou committed:
Apply pyupgrade with correct Python 3.10+ version

- Use --py310-plus flag to match pre-commit configuration
- Apply Python 3.10+ syntax updates
- Ensure compatibility with pre-commit pyupgrade hook

linshenghou committed:
Fix final 6 ruff linting errors

- Apply remaining ruff fixes to all Python files
- Ensure all linting issues are completely resolved
- Final step to achieve full pre-commit compliance

forgetwhatuwant commented:
> nhnifong

i will maybe work on it in soon

nhnifong committed:
Make async encoding work with 3.0 dataset format

nhnifong committed:
Test sync and async episode recording

nhnifong commented:
@forgetwhatuwant I have spend the last day or two pulling this up to dataset 3.0.
https://github.com/nhnifong/lerobot/tree/feature/gpu-async-encoding-speed-enhancement
It is almost working correctly and I am tracking down functions that appear to assume metadata about chunk indices is populated when It might not be.
Please feel free to call me if you have the time to chat about it. ageofabundance on discord

forgetwhatuwant commented:
> * **[torchcodec](https://github.com/pytorch/torchcodec)** - Published PyTorch media encoding library
> * **PyNvVideoCodec** - NVIDIA's Python API for video encoding

Excellent work!
It is a great job of u. And there are some details I want to get in touch, especially which methods to integrate gpu encoding with new datasets format.
Can u tell me more about it, thank u!

nhnifong committed:
Re-enable GPU video encoding feature
Cause workers to wait for all the images in their working directory to be complete.

nhnifong committed:
Update metadata for episode differently when async vid recording is complete

nhnifong committed:
Handle case where episode is resumed with download_videos=False

linshenghou committed:
Merge pull request #1 from nhnifong/feature/gpu-async-encoding-speed-enhancement

Feature/gpu async encoding speed enhancement

linshenghou committed:
Merge branch 'main' into feature/gpu-async-encoding-speed-enhancement

imstevenpmwork commented:
Thank you for kicking this off — it genuinely pushed us to prioritise it. Async/streaming video encoding with hardware (NVENC) codec support has since landed natively in #2974 and #2978: see `StreamingVideoEncoder` in `src/lerobot/datasets/video_utils.py` and the `h264_nvenc`/`hevc_nvenc` options in `src/lerobot/configs/video.py`, exposed as `--dataset.streaming_encoding` / `--dataset.rgb_encoder.vcodec`. Closing as superseded, but thanks again for the push and the benchmarks.
